pydantic/generics.py

```diff
@@ -167,7 +167,12 @@ def replace_types(type_: Any, type_map: Mapping[Any, Any]) -> Any:
             # If all arguments are the same, there is no need to modify the
             # type or create a new object at all
             return type_
-        if origin_type is not None and isinstance(type_, typing_base) and not isinstance(origin_type, typing_base):
+        if (
+            origin_type is not None
+            and isinstance(type_, typing_base)
+            and not isinstance(origin_type, typing_base)
+            and type_._name is not None
```

Comment:
@PrettyWood I can imagine that your work described here: https://github.com/samuelcolvin/pydantic/issues/2465#issuecomment-798766083 would also fix this, but perhaps this is a smaller change and faster fix short term.
In addition, I think adding the tests for user defined generic types is probably a good idea regardless, wdyt?

Reply:
My work was pure R&D don't worry! It was more for reference. Thanks 👍

Reply:
Shouldn't we use `type_.get('_name')`? Are you sure `type_._name` is really always defined? I'm a bit paranoid with `typing` types as there may be some weird hidden type without this `_name` attribute 😅

Reply:
I agree,  changed this